### PR TITLE
Unwrap HandlerFailedException for sync Messenger messages too

### LINKED ISSUE #469: Improve the `Sentry\Integration\IgnoreErrorsIntegration` so it unwraps messenger exceptions?
The `Sentry\Integration\IgnoreErrorsIntegration` operates on the assumption that the first exception in a trace is the one where filtering should occur: https://github.com/getsentry/sentry-php/blob/4f6f8fa701e5db53c04f471b139e7d4f85831f17/src/Integration/IgnoreErrorsIntegration.php#L95

That is not applicable to the context of `symfony/messenger`, because all messages are implicitly wrapped in a `Symfony\Component\Messenger\Exception\HandlerFailedException`, and therefore it is hard to understand whether a failure is to be logged or not.

I wrote a **custom** integration that works around the `HandlerFailedException`, but I'm wondering if this could be made more general-purpose in some way.

Here is the reference code (with tests), so it is possible to use it to improve the bundle, or take ideas from it:

<details>
    <summary>SNIP - code is quite large, so I collapsed it for you</summary>

```php
<?php

declare(strict_types=1);

namespace Core\Infrastructure\Sentry\Integration;

use function array_filter;
use function array_key_exists;
use function array_values;
use Core\Domain\Customer\Exception\InvalidSecretException;
use Core\Domain\Customer\Exception\UserLoggedInException;
use Core\Domain\Exception\DomainException;
use Core\Domain\Exception\ViolationException;
use Core\Infrastructure\Api\Exception\MalformedJsonException;
use Core\Infrastructure\Api\Subscriber\JsonExceptionSubscriber;
use function is_a;
use Sentry\Event;
use Sentry\Integration\IntegrationInterface;
use Sentry\SentrySdk;
use Sentry\State\Scope;
use Symfony\Component\HttpKernel\Exception\HttpException;
use Symfony\Component\Messenger\Exception\HandlerFailedException;
use Symfony\Component\Security\Core\Exception\ExceptionInterface as SymfonySecurityException;

/**
 * We do consider some exceptions "acceptable", because the error is not caused by an
 * infrastructure crash or bug, but rather by invalid data that is supposed to be rejected.
 *
 * Here we:
 *
 *  1. exclude logging of a set of well-known exceptions that we consider "user input error"
 *  2. ensure that we skip logging exceptions of the above category when they are nested in
 *     `symfony/messenger` failure exceptions.
 */
final class IgnoreAcceptableFailuresIntegration implements IntegrationInterface
{
    private const IGNORED_EXCEPTION_TYPES = [
        'rejected constraints on input' => DomainException::class,
        'rejected HTTP requests' => HttpException::class,
        'invalid credentials are sent' => InvalidSecretException::class,
        'input cannot be parsed' => MalformedJsonException::class,
        'rejected/failed authentication' => SymfonySecurityException::class,
        'user authentication is explicitly not wished' => UserLoggedInException::class,
        'input validation operations' => ViolationException::class,
    ];

    /**
     * {@inheritdoc}
     */
    public function setupOnce(): void
    {
        Scope::addGlobalEventProcessor(static function (Event $event): ?Event {
            // Convoluted approach, but emulated from current integration implementations
            $integration = SentrySdk::getCurrentHub()->getIntegration(self::class);

            if (null !== $integration && $integration->shouldDrop($event)) {
                return null;
            }

            return $event;
        });
    }

    private function shouldDrop(Event $event): bool
    {
        $exceptions = array_values($event->getExceptions());

        if ([] === $exceptions) {
            // No exceptions recorded - nothing to filter
            return false;
        }

        $firstExceptionType = $exceptions[0]->getType();

        if (array_key_exists(1, $exceptions)
            && is_a($firstExceptionType, HandlerFailedException::class, true)
        ) {
            $firstExceptionType = $exceptions[1]->getType();
        }

        return [] !== array_filter(
            self::IGNORED_EXCEPTION_TYPES,
            static fn (string $exceptionType): bool => is_a($firstExceptionType, $exceptionType, true)
        );
    }
}
```

```php
<?php

declare(strict_types=1);

namespace App\Tests\Unit\Core\Infrastructure\Sentry\Integration;

use Core\Domain\Customer\Exception\InvalidSecretException;
use Core\Domain\Customer\Exception\UserLoggedInException;
use Core\Domain\Exception\DomainException;
use Core\Domain\Exception\ViolationException;
use Core\Infrastructure\Sentry\Integration\IgnoreAcceptableFailuresIntegration;
use Exception;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;
use Sentry\Event;
use Sentry\ExceptionDataBag;
use Sentry\SentrySdk;
use Sentry\State\HubInterface;
use Sentry\State\Scope;
use stdClass;
use Symfony\Component\HttpKernel\Exception\HttpException;
use Symfony\Component\Messenger\Envelope;
use Symfony\Component\Messenger\Exception\HandlerFailedException;
use Symfony\Component\Security\Core\Exception\ExceptionInterface;

/** @covers \Core\Infrastructure\Sentry\Integration\IgnoreAcceptableFailuresIntegration */
final class IgnoreAcceptableFailuresIntegrationTest extends TestCase
{
    private HubInterface $originalHub;

    /** @var HubInterface&MockObject */
    private HubInterface $hubStub;

    private IgnoreAcceptableFailuresIntegration $integration;

    protected function setUp(): void
    {
        parent::setUp();

        $this->originalHub = SentrySdk::getCurrentHub();
        $this->hubStub = $this->createMock(HubInterface::class);
        $this->integration = new IgnoreAcceptableFailuresIntegration();

        SentrySdk::setCurrentHub($this->hubStub);
    }

    protected function tearDown(): void
    {
        SentrySdk::setCurrentHub($this->originalHub);

        parent::tearDown();
    }

    /**
     * @dataProvider ignoredExceptionTypes
     *
     * @psalm-param class-string<\Throwable> $eventType
     */
    public function testWillDropEventsForWhichTheFirstExceptionTypeTypeIsKnown(string $eventType): void
    {
        $event = Event::createEvent();

        $event->setExceptions([new ExceptionDataBag($this->createStub($eventType))]);

        self::assertNull($this->runIntegration($event));
    }

    public function testWillKeepEventsForWhichTheFirstExceptionTypeTypeIsUnknown(): void
    {
        $event = Event::createEvent();

        $event->setExceptions([new ExceptionDataBag(new Exception())]);

        self::assertSame($event, $this->runIntegration($event));
    }

    /**
     * @dataProvider ignoredExceptionTypes
     *
     * @psalm-param class-string<\Throwable> $eventType
     */
    public function testWillKeepEventsForWhichTheFirstExceptionTypeTypeIsNotKnown(string $eventType): void
    {
        $event = Event::createEvent();

        $event->setExceptions([
            new ExceptionDataBag(new Exception()),
            new ExceptionDataBag($this->createStub($eventType)),
        ]);

        self::assertSame($event, $this->runIntegration($event));
    }

    /**
     * @dataProvider ignoredExceptionTypes
     *
     * @psalm-param class-string<\Throwable> $eventType
     */
    public function testWillDropEventsForWhichTheFirstExceptionTypeTypeIsAMessengerFailureAndTheSecondExceptionIsKnown(string $eventType): void
    {
        $event = Event::createEvent();

        $event->setExceptions([
            new ExceptionDataBag(new HandlerFailedException(
                Envelope::wrap(new stdClass()),
                [new Exception()]
            )),
            new ExceptionDataBag($this->createStub($eventType)),
        ]);

        self::assertNull($this->runIntegration($event));
    }

    public function testWillKeepEventsForWhichTheFirstExceptionTypeTypeIsAMessengerFailureAndTheSecondExceptionIsUnknown(): void
    {
        $event = Event::createEvent();

        $event->setExceptions([
            new ExceptionDataBag(new HandlerFailedException(
                Envelope::wrap(new stdClass()),
                [new Exception()]
            )),
            new ExceptionDataBag(new Exception()),
        ]);

        self::assertSame($event, $this->runIntegration($event));
    }

    public function testWillKeepEventsWithNoExceptions(): void
    {
        $event = Event::createEvent();

        $event->setExceptions([]);

        self::assertSame($event, $this->runIntegration($event));
    }

    private function runIntegration(Event $event): ?Event
    {
        $this->hubStub->method('getIntegration')
            ->willReturnMap([
                [IgnoreAcceptableFailuresIntegration::class, $this->integration],
            ]);

        $this->integration->setupOnce();

        return (new Scope())->applyToEvent($event);
    }

    /** @psalm-return non-empty-list<array{class-string<\Throwable>}> */
    public function ignoredExceptionTypes(): array
    {
        return [
            [DomainException::class],
            [HttpException::class],
            [InvalidSecretException::class],
            [ExceptionInterface::class],
            [UserLoggedInException::class],
            [ViolationException::class],
        ];
    }
}
```
</details>

### PR DESCRIPTION
**Pull request by Jean85**
Will fix #469

### CODE REVIEW

**Review by Ocramius — APPROVED** [review]
**assuming** (needs to be checked by maintainers) that `HandlerFailedException` can be un-wrapped by default within the context of `ErrorListener#handleExceptionEvent()`, I'm totally for this patch :+1:

**Comment by Jean85** [comment]
IMO it should, because it can also contain multiple exceptions, due to multiple handlers for the same message, failing together.

File: tests/End2End/App/Controller/MessengerController.php [hunk 1]
```diff
@@ -33,4 +34,11 @@ public function dispatchUnrecoverableMessage(): Response
 
         return new Response('Success');
     }
+
+    public function dispatchUnrecoverableSyncMessage(): Response
+    {
+        $this->messenger->dispatch(new SyncMessage(false));
```
Comment (ste93cry) [inline tests/End2End/App/Controller/MessengerController.php:40]: The constructor of the `SyncMessage` class does not accept any parameter, not sure why PHPStan does not complain here

File: tests/End2End/App/Messenger/IgnorableException.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sentry\SentryBundle\Tests\End2End\App\Messenger;
+
+class IgnorableException extends \Exception
```
Comment (ste93cry) [inline tests/End2End/App/Messenger/IgnorableException.php:7]: Class can be made `final`

File: tests/End2End/App/Messenger/SyncMessage.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sentry\SentryBundle\Tests\End2End\App\Messenger;
+
+class SyncMessage
```
Comment (ste93cry) [inline tests/End2End/App/Messenger/SyncMessage.php:7]: Class can be made `final`

File: tests/End2End/App/Messenger/SyncMessageHandler.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sentry\SentryBundle\Tests\End2End\App\Messenger;
+
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+class SyncMessageHandler implements MessageHandlerInterface
```
Comment (ste93cry) [inline tests/End2End/App/Messenger/SyncMessageHandler.php:9]: Class can be made `final`

**Review by ste93cry — CHANGES_REQUESTED** [review]
Build is broken 😬

**Comment by Jean85** [comment]
CI failure seems triggered by https://github.com/symfony/symfony/issues/40677

If we cannot fix it, we can (temporarily?) switch to `pcov` for coverage collection?

**Comment by Ocramius** [comment]
@Jean85 IMO using `pcov` is more reliable, if this stuff is only used for coverage.

**Comment by Ocramius** [comment]
```

Failed asserting that actual size 2 matches expected size 1.

/home/runner/work/sentry-symfony/sentry-symfony/tests/End2End/End2EndTest.php:258
/home/runner/work/sentry-symfony/sentry-symfony/tests/End2End/End2EndTest.php:157
```

I wonder if that's because the messenger encounters the fatal error twice?

**Comment by Jean85** [comment]
Why would encounter it twice?

Also, tests do NOT fail locally, there's definitely something wonky here..

**Comment by Ocramius** [comment]
The code around the assertions looks completely wonky:

https://github.com/getsentry/sentry-symfony/blob/43cd505db0a9e370326226467a8a61c7c7deb52f/tests/End2End/End2EndTest.php#L141-L155

The `} catch {` block should not exist there, should it?

**Comment by Jean85** [comment]
IIRC that was done in #322 due to the fact that with different Symfony version you would get different behaviours..

Let's see what happens if I revert that.

**Pull request closed without merging**